src/legacy/core_plugins/kibana/public/dashboard/lib/migrate_app_state.ts

```diff
@@ -17,20 +17,68 @@
  * under the License.
  */
 
-import { SavedDashboardPanel, DashboardAppState } from '../types';
+import semver from 'semver';
+import chrome from 'ui/chrome';
+import { i18n } from '@kbn/i18n';
+import { trackUiMetric } from '../../../../ui_metric/public';
+import {
+  DashboardAppState,
+  SavedDashboardPanelTo60,
+  SavedDashboardPanel730ToLatest,
+  SavedDashboardPanel610,
+  SavedDashboardPanel630,
+  SavedDashboardPanel640To720,
+  SavedDashboardPanel620,
+} from '../types';
+import { migratePanelsTo730 } from '../migrations/migrate_to_730_panels';
 
 /**
- * Creates a new instance of AppState based of the saved dashboard.
+ * Attempts to migrate the state stored in the URL into the latest version of it.
  *
- * @param appState {AppState} AppState class to instantiate
+ * Once we hit a major version, we can remove support for older style URLs and get rid of this logic.
  */
-export function migrateAppState(appState: DashboardAppState) {
-  // For BWC in pre 6.1 versions where uiState was stored at the dashboard level, not at the panel level.
-  if (appState.uiState) {
-    appState.panels.forEach((panel: SavedDashboardPanel) => {
-      panel.embeddableConfig = appState.uiState[`P-${panel.panelIndex}`];
-    });
+export function migrateAppState(appState: { [key: string]: unknown } | DashboardAppState) {
+  if (!appState.panels) {
+    throw new Error(
+      i18n.translate('kbn.dashboard.panel.invalidData', {
+        defaultMessage: 'Invalid data in url',
+      })
+    );
+  }
+
+  const panelNeedsMigration = (appState.panels as Array<
+    | SavedDashboardPanelTo60
+    | SavedDashboardPanel610
+    | SavedDashboardPanel620
+    | SavedDashboardPanel630
+    | SavedDashboardPanel640To720
+    | SavedDashboardPanel730ToLatest
+  >).some(panel => {
+    if ((panel as { version?: string }).version === undefined) return true;
+
+    const version = (panel as SavedDashboardPanel730ToLatest).version;
+
+    // This will help us figure out when to remove support for older style URLs.
+    trackUiMetric('DashboardPanelVersionInUrl', `${version}`);
```

Comment:
Nice! Maybe it is better to move this line after line 67?

Reply:
I think this might be useful information moving forward too, so if at some point we want to introduce a migration say in 8.x, we will know how many 7.x panels are still in use.  Although...  I suppose the panels from 7.x -> 8.x if they didn't go through migrations will all look the same anyway so it shouldn't matter?  So maybe I don't have a concrete reason why this information would be useful but it also seems like, just include it for completeness, because there may be something I'm not thinking about.